app/packages/core/src/components/Modal/modal-spaces-utils.ts

```diff
@@ -0,0 +1,82 @@
+import { PluginComponentRegistration } from "@fiftyone/plugins";
+import { SpaceNodeJSON, usePanels } from "@fiftyone/spaces";
+import { panelsCompareFn } from "@fiftyone/spaces/src/utils/sort";
+import { useCallback, useEffect, useMemo, useRef, useState } from "react";
+
+const MODAL_PLUGINS_REGISRATION_TIMEOUT_MS = 200;
+
+export const SAMPLE_MODAL_PLUGIN_NAME = "fo-sample-modal-plugin";
+const SAMPLE_MODAL_PLUGINS_LOCAL_STORAGE_KEY = "fo-sample-modal-plugins";
+
+export const useModalSpaces = () => {
+  const [modalSpaces, setModalSpaces] = useState<SpaceNodeJSON | null>(null);
+
+  const panelsPredicate = useCallback(
+    (panel: PluginComponentRegistration) =>
+      panel.panelOptions?.surfaces === "modal" ||
+      panel.panelOptions?.surfaces === "grid modal",
+    []
+  );
+
+  const allModalPlugins = usePanels(panelsPredicate);
+
+  const defaultModalSpaces = useMemo(() => {
+    const sortedPlugins = allModalPlugins.sort(panelsCompareFn);
+
+    return {
+      id: "root",
+      children: sortedPlugins.map((modalPlugin) => ({
+        id: `${modalPlugin.name}`,
+        type: modalPlugin.name,
+        pinned: modalPlugin.name === SAMPLE_MODAL_PLUGIN_NAME,
+        children: [],
+      })),
+      type: "panel-container",
+      activeChild: SAMPLE_MODAL_PLUGIN_NAME,
+    } as SpaceNodeJSON;
+  }, [allModalPlugins]);
+
+  const defaultModalSpacesRef = useRef<typeof defaultModalSpaces | null>(null);
+
+  defaultModalSpacesRef.current = defaultModalSpaces;
+
+  useEffect(() => {
+    let timeOutId = -1;
+
+    const maybeModalSpaces = getModalSpacesFromLocalStorage();
+    if (maybeModalSpaces) {
+      setModalSpaces(maybeModalSpaces);
+    } else {
+      // this is a hack to wait for the plugins to be registered
```

Comment:
can we avoid this? Happy to chat offline if needed

Reply:
Happy to explore better alternatives.